File: src/hanlint/rules/shared/candidates.py
"""뜻을 고르지 않고 표층에서 셀 수 있는 후보를 만든다.

후보의 공통 형식은 text와 why 둘뿐이다. 규칙마다 다른 점수나 순위를 두지 않는다. 후보가 없으면 빈
튜플을 주며, 억지로 한 칸을 채우지 않는다.
"""

from __future__ import annotations

import re

from ...analysis.grammar import HAEYO, HANDA, HAPNIDA, REGISTERS, fitJosa, hangul
from ...analysis.grammar.predicate import conjugate, polite
from ...analysis.grammar.voice import decomposePassive
from ...analysis.surface.tokenize import stripJosa, words
from ...data import loadLines
from ...fingerprint import SentencePrint
from ..finding import Candidate

ENDING_CANDIDATES = {
    HAPNIDA: (("-습니까?", "같은 문체의 의문형"), ("-기 때문입니다", "같은 문체의 인과형")),
    HANDA: (("-는가?", "같은 문체의 의문형"), ("-기 때문이다", "같은 문체의 인과형")),
    HAEYO: (("-나요?", "같은 문체의 의문형"), ("-기 때문이에요", "같은 문체의 인과형")),
}


def longSentenceCandidates(text: str) -> tuple[Candidate, ...]:
    """연결 어미 뒤의 문장 경계 후보. 글자를 고쳐 쓰지는 않는다."""
    endings = sorted(loadLines("candidateSplitEndings.txt"), key=len, reverse=True)
    pattern = re.compile(r"(?P<word>[가-힣]+(?P<ending>" + "|".join(map(re.escape, endings)) + r"))[,;]?\s+")
    candidates = []
    for match in pattern.finditer(text):
        word = match.group("word")
        ending = match.group("ending")
        if word == "그리고" or len(word) <= len(ending):
            continue
        marked = text[: match.end()].rstrip() + " | " + text[match.end() :].lstrip()
        candidates.append(Candidate(marked, f"연결 어미 `{ending}` 뒤를 문장 경계로 검토한다"))
    return tuple(candidates)


def _replacementForDeixis(topic: str, deixis: str) -> str:
    if deixis.startswith("이것"):
        suffix = deixis[len("이것") :]
        return topic + fitJosa(topic, suffix)
    if deixis.startswith("해당 "):
        last = deixis.rsplit(maxsplit=1)[-1]
        base = stripJosa(last)
        suffix = last[len(base) :]
        return topic + fitJosa(topic, suffix)
    if deixis == "이러한":
        return topic + "의"
    if deixis == "이처럼":
        return topic + "처럼"
    return topic


def danglingDeixisCandidates(current: SentencePrint, previous: SentencePrint | None) -> tuple[Candidate, ...]:
    """앞 문장에 실제로 나온 화제어를 현재 지시어 자리에 넣은 후보."""
    if previous is None or not current.deixis:
        return ()
    deixis = current.deixis[0]
    topics = sorted(previous.topics, key=lambda topic: (previous.text.lower().find(topic), topic))
    return tuple(
        Candidate(
            current.text.replace(deixis, _replacementForDeixis(topic, deixis), 1),
            f"바로 앞 문장에 나온 명사 `{topic}`",
        )
        for topic in topics
    )


def nounPileCandidates(text: str) -> tuple[Candidate, ...]:
    """명사 나열 안에서 하다 동사로 되돌릴 수 있는 어근."""
    actionNouns = set(loadLines("candidateActionNouns.txt"))
    found = []
    for word in words(text):
        core = stripJosa(word.core)
        if core in actionNouns and core not in found:
            found.append(core)
    return tuple(Candidate(noun + "하다", f"명사 `{noun}`{fitJosa(noun, '을')} 동사 어근으로 되돌린다") for noun in found)


def endingRepeatCandidates(register: str) -> tuple[Candidate, ...]:
    """문서 문체 안에서 고를 수 있는 종결 꼴. 뜻이 맞는지는 고르는 쪽이 정한다."""
    target = register if register in REGISTERS else HAPNIDA
    return tuple(Candidate(text, why) for text, why in ENDING_CANDIDATES[target])


def _reducedPassiveText(text: str, surface: str, reduced: str) -> str | None:
    """외부 `지`의 활용을 첫 피동 어간에 옮긴 문장."""
    prefix = surface[:-1]
    start = 0
    while (at := text.find(prefix, start)) >= 0:
        tailAt = at + len(prefix)
        if tailAt >= len(text) or not hangul.isSyllable(text[tailAt]):
            start = at + 1
            continue
        initial, vowel, final = hangul.split(text[tailAt])
        if initial != 12 or vowel not in (hangul.IH, hangul.YEO):
            start = at + 1
            continue
        if vowel == hangul.YEO:
            base = conjugate(reduced)
            replacement = base[:-1] + hangul.withFinal(base[-1], final) if final else base
        elif final == hangul.BIEUP:
            replacement = polite(reduced)[:-2]
        elif final:
            replacement = reduced[:-1] + hangul.withFinal(reduced[-1], final)
        else:
            replacement = reduced
        return text[:at] + replacement + text[tailAt + 1 :]
    return None


def doublePassiveCandidates(text: str, passives: tuple[str, ...]) -> tuple[Candidate, ...]:
    """이중 피동에서 어지 하나를 뺀 활용 후보."""
    candidates = []
    seen = set()
    for surface in passives:
        voice = decomposePassive(surface)
        if voice is None or voice.reduced is None:
            continue
        candidateText = _reducedPassiveText(text, surface, voice.reduced)
        if candidateText is None or candidateText in seen:
            continue
        seen.add(candidateText)
        candidates.append(Candidate(candidateText, f"`{surface}`의 피동 겹을 하나로 줄인다"))
    return tuple(candidates)

File: npm/src/rules/shared/candidates.js
// @ts-check
/** 파이썬 rules/shared/candidates.py의 투영. 후보에는 점수와 순위가 없다. */
import { HAEYO, HANDA, HAPNIDA, REGISTERS, conjugate, fitJosa } from "../../analysis/grammar/index.js";
import * as hangul from "../../analysis/grammar/hangul.js";
import { decomposePassive } from "../../analysis/grammar/voice.js";
import { stripJosa, words } from "../../analysis/tokenize.js";
import { loadLines } from "../../data/load.js";

const ENDING_CANDIDATES = {
  [HAPNIDA]: [["-습니까?", "같은 문체의 의문형"], ["-기 때문입니다", "같은 문체의 인과형"]],
  [HANDA]: [["-는가?", "같은 문체의 의문형"], ["-기 때문이다", "같은 문체의 인과형"]],
  [HAEYO]: [["-나요?", "같은 문체의 의문형"], ["-기 때문이에요", "같은 문체의 인과형"]],
};

/** @param {string} text @returns {import("../finding.js").Candidate[]} */
export function longSentenceCandidates(text) {
  const endings = [...loadLines("candidateSplitEndings.txt")].sort((a, b) => b.length - a.length);
  const pattern = new RegExp(`([가-힣]+(${endings.join("|")}))[,;]?\\s+`, "g");
  const candidates = [];
  for (const match of text.matchAll(pattern)) {
    const word = match[1];
    const ending = match[2];
    if (word === "그리고" || word.length <= ending.length) continue;
    const end = /** @type {number} */ (match.index) + match[0].length;
    const marked = text.slice(0, end).trimEnd() + " | " + text.slice(end).trimStart();
    candidates.push({ text: marked, why: `연결 어미 \`${ending}\` 뒤를 문장 경계로 검토한다` });
  }
  return candidates;
}

/** @param {string} topic @param {string} deixis */
function replacementForDeixis(topic, deixis) {
  if (deixis.startsWith("이것")) {
    const suffix = deixis.slice("이것".length);
    return topic + fitJosa(topic, suffix);
  }
  if (deixis.startsWith("해당 ")) {
    const last = deixis.trim().split(/\s+/).at(-1) ?? "";
    const base = stripJosa(last);
    return topic + fitJosa(topic, last.slice(base.length));
  }
  if (deixis.startsWith("이러한 ")) return topic + "의 " + deixis.slice("이러한 ".length);
  if (deixis === "이처럼") return topic + "처럼";
  return topic;
}

/** @param {import("../../fingerprint/build.js").SentencePrint} current @param {import("../../fingerprint/build.js").SentencePrint | null} previous */
export function danglingDeixisCandidates(current, previous) {
  if (previous === null || !current.deixis.length) return [];
  const deixis = current.deixis[0];
  const topics = [...previous.topics].sort((a, b) => previous.text.toLowerCase().indexOf(a) - previous.text.toLowerCase().indexOf(b) || a.localeCompare(b));
  return topics.map((topic) => ({
    text: current.text.replace(deixis, replacementForDeixis(topic, deixis)),
    why: `바로 앞 문장에 나온 명사 \`${topic}\``,
  }));
}

/** @param {string} text */
export function nounPileCandidates(text) {
  const actionNouns = new Set(loadLines("candidateActionNouns.txt"));
  const found = [];
  for (const word of words(text)) {
    const core = stripJosa(word.core);
    if (actionNouns.has(core) && !found.includes(core)) found.push(core);
  }
  return found.map((noun) => ({ text: noun + "하다", why: `명사 \`${noun}\`${fitJosa(noun, "을")} 동사 어근으로 되돌린다` }));
}

/** @param {string} register */
export function endingRepeatCandidates(register) {
  const target = REGISTERS.includes(register) ? register : HAPNIDA;
  return ENDING_CANDIDATES[target].map(([text, why]) => ({ text, why }));
}

/** @param {string} text @param {string} surface @param {string} reduced */
function reducedPassiveText(text, surface, reduced) {
  const prefix = surface.slice(0, -1);
  let start = 0;
  while ((start = text.indexOf(prefix, start)) >= 0) {
    const tailAt = start + prefix.length;
    const tail = text[tailAt];
    if (!tail || !hangul.isSyllable(tail)) {
      start += 1;
      continue;
    }
    const [initial, vowel, final] = hangul.split(tail);
    if (initial !== 12 || ![hangul.IH, hangul.YEO].includes(vowel)) {
      start += 1;
      continue;
    }
    let replacement;
    if (vowel === hangul.YEO) {
      const base = conjugate(reduced);
      replacement = final ? base.slice(0, -1) + hangul.withFinal(base.at(-1), final) : base;
    } else if (final === hangul.BIEUP) {
      const polite = renderPolite(reduced);
      replacement = polite.slice(0, -2);
    } else {
      replacement = final ? reduced.slice(0, -1) + hangul.withFinal(reduced.at(-1), final) : reduced;
    }
    return text.slice(0, start) + replacement + text.slice(tailAt + 1);
  }
  return null;
}

/** polite는 공개 함수가 아니므로 합니다체 활용에서 니다만 뗀다. @param {string} reduced */
function renderPolite(reduced) {
  const last = hangul.finalOf(reduced.at(-1));
  if (last === hangul.NONE || last === hangul.RIEUL) {
    const stem = last === hangul.RIEUL ? reduced.slice(0, -1) + hangul.withFinal(reduced.at(-1), hangul.NONE) : reduced;
    return stem.slice(0, -1) + hangul.withFinal(stem.at(-1), hangul.BIEUP) + "니다";
  }
  return reduced + "습니다";
}

/** @param {string} text @param {string[]} passives */
export function doublePassiveCandidates(text, passives) {
  const candidates = [];
  const seen = new Set();
  for (const surface of passives) {
    const voice = decomposePassive(surface);
    if (!voice || voice.reduced === null) continue;
    const candidateText = reducedPassiveText(text, surface, voice.reduced);
    if (candidateText === null || seen.has(candidateText)) continue;
    seen.add(candidateText);
    candidates.push({ text: candidateText, why: `\`${surface}\`의 피동 겹을 하나로 줄인다` });
  }
  return candidates;
}
